### PR TITLE
Query: Relational: Second round of major SQL improvements.

### PR DESCRIPTION
**Pull request by anpete**
- Enables projection optimization across client-eval'd subquery boundaries. Subqueries were previously fully materialized.
- Subqueries in AdditionalFromClauses are now lifted where possible.

### CODE REVIEW

**Comment by anpete** [comment]
@maumar

File: src/EntityFramework.Core/Query/ExpressionVisitors/RequiresMaterializationExpressionVisitor.cs [hunk 1]
```diff
@@ -47,6 +62,8 @@ public virtual ISet<IQuerySource> QuerySourcesRequiringMaterialization
 
         protected override Expression VisitMember(MemberExpression memberExpression)
         {
+            Check.NotNull(memberExpression, nameof(memberExpression));
+
             var newExpression = base.VisitMember(memberExpression);
```
Comment (maumar) [inline src/EntityFramework.Core/Query/ExpressionVisitors/RequiresMaterializationExpressionVisitor.cs]: consider refactoring to Visit(memberExpression.Expression)

File: src/EntityFramework.Core/Query/ExpressionVisitors/MemberAccessBindingExpressionVisitor.cs [hunk 2]
```diff
@@ -36,20 +34,57 @@ public class MemberAccessBindingExpressionVisitor : ReferenceReplacingExpression
             _inProjection = inProjection;
         }
 
+        protected override Expression VisitNew(NewExpression newExpression)
+        {
+            Check.NotNull(newExpression, nameof(newExpression));
+
+            var newArguments = Visit(newExpression.Arguments).ToList();
```
Comment (maumar) [inline src/EntityFramework.Core/Query/ExpressionVisitors/MemberAccessBindingExpressionVisitor.cs]: I wonder if we can avoid rebuilding this array every time (and hence creating new instance of the NewExpression, even if we dont have to)

**Comment by maumar** [comment]
:shipit:

**Pull request closed without merging**